xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,240 @@
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  const auto& addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+
+  if (base)
+  {
+    return base->GetActiveAddon();
+  }
+
+  return nullptr;
+}
+
+void CBinaryAddonManager::OnEvent(const AddonEvent& event)
+{
+  if (auto enableEvent = dynamic_cast<const AddonEvents::Enabled*>(&event))
+  {
+    EnableEvent(enableEvent->id);
+  }
+  else if (auto disableEvent = dynamic_cast<const AddonEvents::Disabled*>(&event))
+  {
+    DisableEvent(disableEvent->id);
+  }
+  else if (typeid(event) == typeid(AddonEvents::InstalledChanged))
+  {
+    InstalledChangeEvent();
+  }
+}
+
+void CBinaryAddonManager::EnableEvent(const std::string& addonId)
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  const auto& addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+  else
+    return;
+
+  CLog::Log(LOGDEBUG, "Binary addons: Enable addon '%s' on binary addon manager", base->ID().c_str());
+  m_enabledAddons[base->ID()] = base;
+
+  /**
+   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
+   *
+   * Currently only Screensaver and Visualization use the new way and not need informed.
+   */
+}
+
+void CBinaryAddonManager::DisableEvent(const std::string& addonId)
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  const auto& addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+  else
+    return;
+
+  CLog::Log(LOGDEBUG, "Binary addons: Disable addon '%s' on binary addon manager", base->ID().c_str());
+  m_enabledAddons.erase(base->ID());
+
+  /**
+   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
+   *
+   * Currently only Screensaver and Visualization use the new way and not need informed.
+   */
+}
+
+void CBinaryAddonManager::InstalledChangeEvent()
+{
+  CSingleLock lock(m_critSection);
+
+  std::vector<std::tuple<std::string, bool, std::string>> binaryAddonList;
+  CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList);
+
+  BinaryAddonMgrBaseList deletedAddons = m_installedAddons;
+  for (auto addon : binaryAddonList)
+  {
+    const auto& knownAddon = m_installedAddons.find(std::get<2>(addon));
+    if (knownAddon == m_installedAddons.end())
+    {
+      CLog::Log(LOGDEBUG, "Binary addons: Adding new binary addon '%s'", std::get<2>(addon).c_str());
+
+      BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>();
+      if (!base->Create(std::get<0>(addon)))
+      {
+        CLog::Log(LOGERROR, "Binary Addons: Failed to create base for '%s' and addon not usable", std::get<0>(addon).c_str());
+        continue;
+      }
+      m_installedAddons[base->ID()] = base;
+      if (std::get<1>(addon))
+        m_enabledAddons[base->ID()] = base;
```

Comment:
at the first glance, the above code lines look pretty much the same as found in <code>bool CBinaryAddonManager::Init()</code>. maybe you should factor out the identical code?